modules/ismp/testsuite/src/lib.rs

```diff
@@ -402,7 +429,128 @@ pub fn check_request_source_and_destination() {}
 
 /// This should check that if a proxy isn't configured, responses are not valid if they don't come
 /// from the state machine claimed in the proof
-pub fn check_response_source() {}
+pub fn check_response_source<H, D>(host: &H, dispatcher: &D) -> Result<(), &'static str>
+where
+    H: IsmpHost,
+    D: IsmpDispatcher,
+    D::Account: From<[u8; 32]>,
+    D::Balance: From<u32>,
+{
+    let intermediate_state = setup_mock_client(host);
+    // Set the previous update time
+    let challenge_period = host.challenge_period(mock_consensus_state_id()).unwrap();
+    let previous_update_time = host.timestamp() - (challenge_period * 2);
+    host.store_consensus_update_time(mock_consensus_state_id(), previous_update_time)
+        .unwrap();
+    host.store_state_machine_update_time(intermediate_state.height, previous_update_time)
+        .unwrap();
+
+    assert!(host.allowed_proxy().is_none());
+
+    let post = Post {
+        source: intermediate_state.height.id.state_id,
+        dest: host.host_state_machine(),
+        nonce: 0,
+        from: vec![0u8; 32],
+        to: vec![0u8; 32],
+        timeout_timestamp: 0,
+        data: vec![0u8; 64],
+        gas_limit: 0,
+    };
+
+    // Request message handling check for source and destination chain
+    let request_message = Message::Request(RequestMessage {
+        requests: vec![post.clone()],
+        proof: Proof { height: intermediate_state.height, proof: vec![] },
+        signer: vec![0u8; 32],
+    });
+
+    handle_incoming_message(host, request_message).unwrap();
+    // Assert that request was acknowledged
+    assert!(host.request_receipt(&Request::Post(post)).is_some());
+
+    let post = Post {
+        source: StateMachine::Kusama(2000),
+        dest: host.host_state_machine(),
+        nonce: 0,
+        from: vec![0u8; 32],
+        to: vec![0u8; 32],
+        timeout_timestamp: 0,
+        data: vec![0u8; 64],
+        gas_limit: 0,
+    };
+    let response =
+        PostResponse { post: post.clone(), response: vec![], timeout_timestamp: 0, gas_limit: 0 };
+    dispatcher
+        .dispatch_response(response.clone(), [0; 32].into(), 0u32.into())
```

Comment:
We want to test handling of a post response message not dispatching a post response. The test should try to handle a post response where the source of the response does not match the proof source.